cmd/thanos/rule.go

```diff
@@ -459,7 +463,23 @@ func runRule(
 			close(cancel)
 		})
 	}
-
+	// Periodically update the addresses from static flags and file SD by resolving them using DNS SD if necessary.
+	{
+		ctx, cancel := context.WithCancel(context.Background())
+		g.Add(func() error {
+			return runutil.Repeat(dnsSDInterval, ctx.Done(), func() error {
+				addresses := append(fileSDCache.Addresses(), queryAddrs...)
+				// TODO(ivan): default port... Use a flag maybe?
+				if err := dnsProvider.Resolve(ctx, addresses, 9090); err != nil {
+					// Failure to resolve could be caused by a lookup timeout. We shouldn't fail because of that, so just log.
+					level.Warn(logger).Log("msg", fmt.Sprintf("failed to resolve addresses in query: %v", err.Error()))
```

Comment:
Sounds like error to me. Plus some metric?

Reply:
I chose Warning because the DNS discovery used for alertmanagers used Warning when it failed, but I also think it looks more like an error. I will also add metrics.